src/tbb/waiters.h

```diff
@@ -50,14 +50,41 @@ class waiter_base {
 };
 
 class outermost_worker_waiter : public waiter_base {
+    friend class arena;
 public:
     using waiter_base::waiter_base;
 
-    bool continue_execution(arena_slot& slot, d1::task*& t) const {
+    bool continue_execution(arena_slot& slot, d1::task*& t) {
         __TBB_ASSERT(t == nullptr, nullptr);
 
         if (is_worker_should_leave(slot)) {
-            // Leave dispatch loop
+            market* m = my_arena.my_market;
+
+            int next_epoch{};
+            int current_epoch = m->my_adjust_demand_current_epoch.load(std::memory_order_relaxed);
+            bool is_same_epoch = false;
+
+            auto t1 = std::chrono::steady_clock::now();
+            for (auto t2 = std::chrono::steady_clock::now();
+                std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1) < std::chrono::microseconds(40);
+                t2 = std::chrono::steady_clock::now())
+            {
+                if (!is_same_epoch) {
+                    arena* a = m->arena_in_need(&my_arena, &my_arena);
```

Comment:
I would check `a.is_recall_requested` and skip the current arena in `arena_in_need` instead of duplication of `is_recall_requested` logic in `arena_in_need`